mindsdb/interfaces/database/data_handlers_cache.py

```diff
@@ -0,0 +1,190 @@
+        if isinstance(self.cleaner_thread, threading.Thread) and self.cleaner_thread.is_alive():
+            return
+        with self._lock:
+            self._stop_event.clear()
+            self.cleaner_thread = threading.Thread(target=self._clean, name="HandlersCache.clean")
+            self.cleaner_thread.daemon = True
+            self.cleaner_thread.start()
+
+    def _stop_clean(self) -> None:
+        """stop clean worker"""
+        self._stop_event.set()
+
+    def set(self, handler: DatabaseHandler):
+        """add (or replace) handler in cache
+
+        NOTE: If the handler is not thread-safe, then use a lock when making connection. Otherwise, make connection in
+        the same thread without using a lock to speed up parallel queries. (They don't need to wait for a connection in
+        another thread.)
+
+        Args:
+            handler (DatabaseHandler)
+        """
+        thread_safe = getattr(handler, "thread_safe", True)
+        with self._lock:
+            try:
+                # If the handler is defined to be thread safe, set 0 as the last element of the key, otherwise set the thrad ID.
+                key = (
+                    handler.name,
+                    ctx.company_id,
+                    0 if thread_safe else threading.get_native_id(),
+                )
+                record = HandlersCacheRecord(handler=handler, expired_at=time.time() + self.ttl)
+                self.handlers[key].append(record)
+            except Exception:
+                logger.warning("Error setting data handler cache record:", exc_info=True)
+                return
+        self._start_clean()
+        record.connect()
+
+    def _get_cache_records(self, name: str) -> tuple[list[HandlersCacheRecord] | None, str]:
+        """get cache records by name
+
+        Args:
+            name (str): handler name
+
+        Returns:
+            tuple[list[HandlersCacheRecord] | None, str]: cache records and key of the handler in cache
+        """
+        # If the handler is not thread safe, the thread ID will be assigned to the last element of the key.
+        key = (name, ctx.company_id, 0)
+        if key not in self.handlers:
+            key = (name, ctx.company_id, threading.get_native_id())
+        return self.handlers.get(key, []), key
+
+    def get(self, name: str) -> DatabaseHandler | None:
+        """get handler from cache by name
+
+        Args:
+            name (str): handler name
+
+        Returns:
+            DatabaseHandler
+        """
+        with self._lock:
+            records, _ = self._get_cache_records(name)
+            for record in records:
+                if record.expired is False and record.has_references is False:
+                    record.expired_at = time.time() + self.ttl
+                    if record.connect_attempt_done.wait(timeout=10) is False:
+                        logger.warning(f"Handler's connection attempt has not finished in 10s: {record.handler.name}")
```

Comment:
atually there is no connection attempt here. it will wait 10s every time if something went wrong with connection in `set`